modules/dnn/src/onnx/onnx_importer.cpp

```diff
@@ -3260,6 +3283,74 @@ void ONNXImporter::parseQMatMul(LayerParams& layerParams, const opencv_onnx::Nod
     addLayer(layerParams, node_proto);
 }
 
+// A * B + C = Y, we require that the dimension of A is [m, k], and the dimension of B is [n, k].
+// And the dim of output Y is [m, n]
+void ONNXImporter::parseQGemm(LayerParams& layerParams, const opencv_onnx::NodeProto& node_proto)
+{
+    int ninputs = node_proto.input_size();
+    CV_Assert(ninputs == 8 || ninputs == 9);
+
+    layerParams.type = "InnerProductInt8";
+
+    if (constBlobs.find(node_proto.input(3)) == constBlobs.end())
+        CV_Error(Error::StsNotImplemented, "Variable weights is not supported");
+
+    Mat weights = getBlob(node_proto, 3);
+
+    if (!layerParams.get<int>("transB", 0))
+    {
+        transpose(weights, weights);
+    }
+
+    int firstInpDims = outShapes[node_proto.input(0)].size();
+
+    Mat inp_sc = getBlob(node_proto, 1);
+    Mat inp_zp = getBlob(node_proto, 2);
+
+    int outCn = weights.size[0];
+    int secondInpDims = weights.dims;
+
+    Mat w_scale = getBlob(node_proto, 4);
+    CV_Assert(w_scale.total() == 1 || w_scale.total() == outCn);
+    bool per_channel = w_scale.total() == outCn ? true : false;
+    Mat wt_sc = (w_scale.total() == outCn) ? w_scale : Mat(1, outCn, CV_32F, Scalar(w_scale.at<float>(0)));
+
+    Mat out_sc, bias;
+    if (ninputs == 9)
+    {
+        out_sc = getBlob(node_proto, 7);
+        if (constBlobs.find(node_proto.input(6)) != constBlobs.end())
+            bias = getBlob(node_proto, 6);
+        else
+            bias = Mat::zeros(1, outCn, CV_32S);
+    }
+    else // ninputs == 8
+    {
+        out_sc = getBlob(node_proto, 6);
+        bias = Mat::zeros(1, outCn, CV_32S);
+    }
```

Comment:
Again, we assume that input `C` is skipped, but that might not be the case. And I think there should be "7" either way - the sixth tensor is just empty, no?

Reply:
Currently, I can only support 8 or 9 inputs.
If it has 9 inputs, there are two different cases: the sixth tensor is empty or not.
If it has 8 inputs, we think the `C` is empty.
This is not a complete solution, but it will handle most cases.

Reply:
I think we should test  the "else" case, because as far as I understand [this](https://github.com/onnx/onnx/blob/main/docs/IR.md#optional-inputs-and-outputs) paragraph, out_sc should still be the seventh blob. Either way, if we expect something to be missing - we should add an assertion for that.

In case of 8 inputs I think we don't have `y_zero_point`, and in case of 9 it seems that we ignore it even if we have it.

Reply:
Agree, now I get your points.

Reply:
Thank you! I see that `C` and `y_scale` tensors are being handled correctly now. But what about `y_zero_point`? I think we should assert that either `ninputs == 8` or `node_proto.input(8).empty()`, or `that 8'th blob contains scalar zero`.

Reply:
In my understanding, the current `QGemm` layer does not need to care about the `output_zeropoint`. The `input_zeropoint` of his consumer is equal to the output_zeropoint of `QGemm`. We can process it when we get `input_zeropoint` in the next layer.
In `QConv`, `QMatMul` and all other quantization layers, we only set the `input_scale` and `input_zeropoint` when we create a new quantized layer.

Reply:
I don't see why would output zero point be so different from output scale - we care about the latter, but not the former?

Reply:
Hi @rogday, I have checked the detail of the quantized structure in DNN. It was my fault, we have uniformly set the `output_scale` and `output_zeropoint` [here](https://github.com/opencv/opencv/blob/4.x/modules/dnn/src/onnx/onnx_importer.cpp#L756-L799). So we don't need to set them again during the node parsing stage. And in some cases, we need to calculate the `outputMultiplier`, that's why we need to get the `output_scale`.

Reply:
It seems unreasonable to calculate the `outputMultiplier` in the parse stage, maybe we can move it to the initialization stage of every quantized node.

Reply:
That function seems very fragile. It relies on the names the producer decided, not the official spec. I think your PR can be merged, but we need to fix this function in the future.

Reply:
Opened #22442

Reply:
Yes, I agree.